lib/PDF/PDFWrapper.php

```diff
@@ -0,0 +1,175 @@
+<?php
+/**
+ * SugarCRM Community Edition is a customer relationship management program developed by
+ * SugarCRM, Inc. Copyright (C) 2004-2013 SugarCRM Inc.
+ *
+ * SuiteCRM is an extension to SugarCRM Community Edition developed by SalesAgility Ltd.
+ * Copyright (C) 2011 - 2021 SalesAgility Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Affero General Public License version 3 as published by the
+ * Free Software Foundation with the addition of the following permission added
+ * to Section 15 as permitted in Section 7(a): FOR ANY PART OF THE COVERED WORK
+ * IN WHICH THE COPYRIGHT IS OWNED BY SUGARCRM, SUGARCRM DISCLAIMS THE WARRANTY
+ * OF NON INFRINGEMENT OF THIRD PARTY RIGHTS.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Affero General Public License along with
+ * this program; if not, see http://www.gnu.org/licenses or write to the Free
+ * Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301 USA.
+ *
+ * You can contact SugarCRM, Inc. headquarters at 10050 North Wolfe Road,
+ * SW2-130, Cupertino, CA 95014, USA. or at email address contact@sugarcrm.com.
+ *
+ * The interactive user interfaces in modified source and object code versions
+ * of this program must display Appropriate Legal Notices, as required under
+ * Section 5 of the GNU Affero General Public License version 3.
+ *
+ * In accordance with Section 7(b) of the GNU Affero General Public License version 3,
+ * these Appropriate Legal Notices must retain the display of the "Powered by
+ * SugarCRM" logo and "Supercharged by SuiteCRM" logo. If the display of the logos is not
+ * reasonably feasible for technical reasons, the Appropriate Legal Notices must
+ * display the words "Powered by SugarCRM" and "Supercharged by SuiteCRM".
+ */
+
+namespace SuiteCRM\PDF;
+
+use SuiteCRM\PDF\Exceptions\PDFEngineNotFoundException;
+use SuiteCRM\PDF\MPDF\MPDFEngine;
+
+if (!defined('sugarEntry') || !sugarEntry) {
+    die('Not A Valid Entry Point');
+}
+
+/**
+ * Class PDFWrapper
+ * @package SuiteCRM\PDF
+ */
+class PDFWrapper
+{
+    /**
+     * @var array stores an associative array matching the PDF engine class name with the file it is stored in.
+     */
+    private static $engines = [
+        'MPDFEngine' => [
+            'name' => 'MPDFEngine',
+            'FQN' => MPDFEngine::class,
+            'filepath' => 'lib/PDF/MPDF/MPDFEngine.php'
+        ],
+    ];
+
+    /** @var string Path to the folder where to load custom engines from */
+    private static $customEnginePath = __DIR__ . '/../../custom/Extension/PDFEngines/';
```

Comment:
Hi @Dillon-Brown does this work similar to the regular extension framework? what configuration is needed to add a new engine?

Reply:
So at the moment, everything in the PDFWrapper is hardcoded to use MPDFEngine. The intention here being that later on we will be able to specify an engine in the ```config.php``` that users can set through the front-end.

For the extension framework, you can just drop a new engine file inside ```custom/Extension/PDFEngines``` and the wrapper will pick it up as an option. I've kept this pretty much exactly the same as the GlobalSearch ```SearchWrapper``` (So much so that I'm thinking of extracting this out further into a common ```WrapperInterface``` for everything inside of ```lib/```).

Reply:
ok thank you